lib/fit/file/data.rb

```diff
@@ -31,34 +51,44 @@ def record_type
                          end
 
             # in case the field size is a multiple of the field length, we must build an array
-             if (field.type != "string" and field.size > field.length)
-               # apply the scale on array only if it has to be applied
-               scale_type = :array if scale_type == :apply
-               code << "array :#{field.raw_name}, :type => :#{field.type}, :initial_length => #{field.size/field.length}\n"
-             else
-               # string are not null terminated when they have exactly the lenght of the field
-               code << "#{field.type} :#{field.raw_name}"
-               if field.type == "string"
-                 scale_type = :string
-                 code << ", :read_length => #{field.size}, :trim_padding => true"
-               end
-               code << "\n"
-             end
+            if (field.type != "string" and field.size > field.length)
+              # apply the scale on array only if it has to be applied
+              scale_type = :array if scale_type == :apply
+              code << "array :#{field.raw_name}, :type => :#{field.type}, :initial_length => #{field.size/field.length}\n"
+            else
+              # string are not null terminated when they have exactly the lenght of the field
+              code << "#{field.type} :#{field.raw_name}"
+              if field.type == "string"
+                scale_type = :string
+                code << ", :read_length => #{field.size}, :trim_padding => true"
+              end
+              code << "\n"
+            end
 
-             code << "def #{field.name}\n"
+            code << "def #{field.name}\n"
 
-             case scale_type
-             when :apply
-               code << "#{field.raw_name}.snapshot / #{field.scale.inspect}.0\n"
-             when :array
-               code << "#{field.raw_name}.snapshot.map { |elt| elt / #{field.scale.inspect}.0 }\n"
-             else
-               code << "#{field.raw_name}.snapshot\n"
-             end
+            if field.dyn_data
+              code << <<-RUBY
+                dyn = #{field.dyn_data}
+                val = get_dyn_value dyn, #{field.raw_name}
+                return val if val
+              RUBY
+              #
+            end
+            case scale_type
+            when :apply
+              code << "#{field.raw_name}.snapshot / #{field.scale.inspect}.0\n"
+            when :array
+              code << "#{field.raw_name}.snapshot.map { |elt| elt / #{field.scale.inspect}.0 }\n"
+            else
+              code << <<-RUBY
+                get_real_value '#{field.real_type}', #{field.raw_name}.snapshot
+              RUBY
+            end
 
-             code << "end\n"
+            code << "end\n"
 
-             class_eval code , __FILE__, __LINE__ + 1
+            class_eval code , __FILE__, __LINE__ + 1
```

Comment:
It feels like this dynamically generated code is getting harder and harder to follow. I wonder if we moved the bindata field, and reader method generation to somewhere else, like inside separate field classes, it would clean things up? Any other ideas?